### ISSUE TITLE
Spring Cosmos - Support patch/partial update

### ISSUE THREAD

**Issue by z1069867141** [Azure/azure-sdk-for-java#21433]
cosmos is a document DB, and one record could be large. Sometimes it's desired to only update part of the document instead of the whole document.

This story is to support the partial patch update to one cosmos document.

**Comment by kushagraThapar** [comment]
Cosmos is waiting for Bulk / Batch / Patch GA, once that is done, this can be implemented in spring.

**Comment by Manuel-Moya** [comment]
Patch is GA :)

**Comment by kushagraThapar** [comment]
@Manuel-Moya - yes, it is GA now, will be released in Java SDK v4.21.0 sometime next week. Once that is out, we will start building support for it in spring-data-cosmos SDK.

**Comment by MihaiTudorP** [comment]
Hello, is there any update on this? It's been some time since the 3rd of November, 2021, with nothing here. Any roadmap milestones or estimated release in which this will come out? The Cosmos feature has been available for quite some time now.

**Comment by backwind1233** [comment]
Hi @kushagraThapar, could you help to give some updates to MihaiTudorP?

**Comment by kushagraThapar** [comment]
@MihaiTudorP - apologies for the missed update. We do plan to pick up Patch API work in this semester and plan to deliver it early next year after the holidays. 
Since we are still looking into the design phase, I am wondering if we can collect some design input from you. And the most important question being how do you plan to use Patch API in spring data cosmos SDK? Since spring-data abstracts out most of the database components, exposing Patch API is going to be interesting from design point of view. 
cc @TheovanKraay 
@trande4884, can you please take this discussion forward?

**Comment by MihaiTudorP** [comment]
> @MihaiTudorP - apologies for the missed update. We do plan to pick up Patch API work in this semester and plan to deliver it early next year after the holidays. Since we are still looking into the design phase, I am wondering if we can collect some design input from you. And the most important question being how do you plan to use Patch API in spring data cosmos SDK? Since spring-data abstracts out most of the database components, exposing Patch API is going to be interesting from design point of view. cc @TheovanKraay @trande4884, can you please take this discussion forward?

Hi, @kushagraThapar , thank you very much for the swift update. Do you have also an estimated date of delivery for this, so that I can plan the version upgrade on my project?

**Comment by trande4884** [comment]
@MihaiTudorP we do not yet have an estimated date of delivery, I will update here once we do

**Comment by TheovanKraay** [comment]
@MihaiTudorP as Kushagra mention, we would also really like to collect some design input from you. 

The most important question being: what is your motivation for wanting to use patch api in Spring? 

If, for example, the motivation is to save on RU costs from doing replace(), keep in mind that the RU cost for patch() is equivalent to the RU for replace() because the physics in terms of physical resources (cpu, memory, etc.) are roughly the same. The cost savings for patch() compared to replace() will come from avoiding the RU cost for a read() and cpu/memory burned on the application that comes before replace(). For example, if I want to update a property – I can either:

1. Send a patch()

Or

2. read() the document + patch within the application tier + replace() the document

Typically Spring users are not needing to do #2 and so there is no benefit in this space. However, another major benefit for patch is avoiding etag contention from highly concurrent updates. 

hence, any feedback on your motivation and use case for using patch, and how you expect it to benefit you, will be greatly appreciated! Thank you.

**Comment by MihaiTudorP** [comment]
@TheovanKraay my motivation is to improve overall application performance (sending less data should improve processing speeds) and avoid etag contention from concurrent updates. Also, less time on the network for sensitive data.

**Comment by arunim29** [comment]
@TheovanKraay @kushagraThapar is there a way to use vanilla cosmos client provided by cosmos sdk along with spring data cosmos. Basically, share the same cosmosClient instance. With this, we can use the best of both worlds.

I could not find a way because spring data cosmos expects a bean of cosmosClientBuilder instead of cosmosClient. If it would have been cosmosClient, I could have shared the instance. Any idea on how to achieve this?

**Comment by TheovanKraay** [comment]
> @TheovanKraay my motivation is to improve overall application performance (sending less data should improve processing speeds) and avoid etag contention from concurrent updates. Also, less time on the network for sensitive data.

Hello @MihaiTudorP, in order to help with design of this, we have a few follow up questions:

1. Is your expectation to use patch via CosmosTemplate or the Repositories?
2. Are you planning to use Reactive Spring or Spring?

**Comment by arunim29** [comment]
@TheovanKraay thanks for a quick reply. I have some follow-up questions :

Any reason why cosmosclient is made private? I am asking because the builder is provided anyways by the application. So, you get all the config from the application. Correct me if I am wrong but in spring data JPA for relation dbs, spring data takes datasource bean from application and that bean can be used in parallel by the application code as well.

**Comment by TheovanKraay** [comment]
> @TheovanKraay thanks for a quick reply. I have some follow-up questions :
> 
> Any reason why cosmosclient is made private? I am asking because the builder is provided anyways by the application. So, you get all the config from the application. Correct me if I am wrong -> in spring data for relation dbs, spring data takes datasource bean from application and that bean can be used in parallel by the application code as well.

Corrected myself. You can access the client directly. See above.

**Comment by arunim29** [comment]
@TheovanKraay if u do not create bean of CosmosClient, which cosmos client do you use to support normal crud repositories which are not reactive.

**Comment by TheovanKraay** [comment]
@arunim29 non-reactive Repository still uses CosmosAsyncClient behind the scenes, but with blocking calls. CosmosClient is not used by Spring Data Cosmos library.

**Comment by arunim29** [comment]
@TheovanKraay got it. Thanks for the help.

**Comment by TheovanKraay** [comment]
@arunim29 - as posted in separate issue #32390 - we plan to expose patch something like the below. Let us know if this is not what you expect. Feel free to make suggestions or ask questions here.

```java
CosmosPatchOperations patchOperations = CosmosPatchOperations
    .create()
    .replace("/size", 5)
    .add("/color", "blue");

User patchedUser = userRepository.patch(user.getId(), new PartitionKey(user.getPartitionKeyField()),patchOperations, User.class);
```
